sway-lsp/src/server.rs

```diff
@@ -124,15 +115,55 @@ fn capabilities() -> ServerCapabilities {
 }
 
 impl Backend {
+    fn url_to_session(&self, uri: &Url) -> Result<Arc<Session>, LanguageServerError> {
+        let path = PathBuf::from(uri.path());
+        let manifest = PackageManifestFile::from_dir(&path).map_err(|_| {
+            DocumentError::ManifestFileNotFound {
+                dir: path.to_string_lossy().to_string(),
+            }
+        })?;
+
+        // strip Forc.toml from the path to get the manifest directory
+        let manifest_dir = manifest
+            .path()
+            .parent()
+            .ok_or(LanguageServerError::ManifestDirNotFound)?
+            .to_path_buf();
+
+        let session = match self.sessions.get(&manifest_dir) {
+            Some(item) => item.value().clone(),
+            None => {
+                // TODO, remove this once the type engine no longer uses global memory: https://github.com/FuelLabs/sway/issues/2063
+                // Until then, we clear the current project and init with the new project.
+                // At least allows the user to switch between projects within a workspace but only if they
+                // have one pane open.
+                let _ = self.sessions.iter().map(|item| {
+                    let session = item.value();
+                    session.shutdown();
+                });
+
+                // If no session can be found, then we need to call init and inserst a new session into the map
+                self.init(uri)?;
+                self.sessions
+                    .get(&manifest_dir) //
+                    .map(|item| item.value().clone())
+                    .expect("no session found even though it was just inserted into the map")
+            }
```

Comment:
It looks like the previous session is not removed from `sessions`, which means if a user switches back the previous session will no longer be active? I guess this is fine for now as the real solution is addressing https://github.com/FuelLabs/sway/issues/2063 so you can actually have multiple simultaneous sessions.

Reply:
Good spot, i'm going to just clear the map after shutting down the session in the meantime. Hopefully we can remove this block of code once #2063 lands and we can store multiple sessions.